### PR TITLE
jobs: add support for SHOW CREATE SCHEDULE command

### LINKED ISSUE #58372: jobs: SHOW CREATE SCHEDULE command
Once the scheduler in enabled, `SHOW SCHEDULES` shows data in a tabular view.  We should have a command such as `SHOW CREATE SCHEDULE` to show the `CREATE SCHEDULE` SQL statement that was used to create the existing schedules.  

Epic CRDB-7909

### PR DESCRIPTION
**Pull request by annezhu98**
Before this change, there was no command to show SQL statements used to create scheduled jobs. This commit allows users to view create statements for scheduled jobs. There are currently two options for this command: 

- `SHOW CREATE SCHEDULE <schedule_id>`: show the create statement for a scheduled job.
- `SHOW CREATE ALL SCHEDULES`: show the create statements for all scheduled jobs.

Example usage:

```
 > SHOW CREATE SCHEDULE 123;
      schedule_id     |                                                   create_statement
 ---------------------+------------------------------------------------------------------------------------------------------------------------
        123           | CREATE SCHEDULE 'core_schedule_label' FOR BACKUP DATABASE defaultdb INTO 'nodelocal://1/my_backup' RECURRING '@daily'
 (1 row)
```

Example execution results:

image
image (1)
image (2)
image

Resolves: #58372

Release note (sql change): added SHOW CREATE SCHEDULES command to view SQL statements used to create existing schedules

### CODE REVIEW

**Review by miretskiy — CHANGES_REQUESTED** [review]
you need to update `pkg/sql/parser/parse_test.go`

> Reviewed 5 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-:-Mcu4Nu94Yic_1FxUWrb:b-yafbrq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @annezhu98)

___
*[pkg/ccl/backupccl/backup.proto, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mcu4t5i2_qX8OtSR0Q8:-Mcu4t5i2_qX8OtSR0Q9:b-dtzg10) ([raw file](https://github.com/cockroachdb/cockroach/blob/9161773c608b8639880a8f8eeba7f3f84f48421d/pkg/ccl/backupccl/backup.proto#L160)):*
> ```protobuf
>   int64 unpause_on_success = 3;
>   bool updates_last_backup_metric = 4;
>   string create_schedule_statement = 5;
> ```

I'm not sure we should store schedule creation command as part of backup execution args.
If this is schedule generic functionality, that should probably go inside
schedule.proto:ExecutionArgs.

?

**Review by annezhu98 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @miretskiy)

___
*[pkg/ccl/backupccl/backup.proto, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mcu4t5i2_qX8OtSR0Q8:-Mcul7GH92neor8nRxkb:b-9gi8o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9161773c608b8639880a8f8eeba7f3f84f48421d/pkg/ccl/backupccl/backup.proto#L160)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not sure we should store schedule creation command as part of backup execution args.
If this is schedule generic functionality, that should probably go inside
schedule.proto:ExecutionArgs.

?
</blockquote></details>

I believe this would be specific to scheduled BACKUPs, similar to `SHOW SCHEDULE`

**Comment by adityamaru** [comment]
> I believe this would be specific to scheduled BACKUPs, similar to `SHOW SCHEDULE`

We might have other scheduled jobs in the future, egs: scheduled exports. So if we place this in a more generic proto, then `SHOW CREATE SCHEDULE` doesn't have to be tied to only scheduled backups, but can show us the SQL query for any scheduled job.

**Review by adityamaru — CHANGES_REQUESTED** [review]
> Reviewed 1 of 9 files at r1, 1 of 1 files at r2, 2 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annezhu98 and @miretskiy)

___
*[pkg/jobs/jobspb/schedule.proto, line 53 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mcy7IPh2oI9Tg2Egyzk:-Mcy7IPh2oI9Tg2Egyzl:bmo4tsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6dfc3bd22b66710435adab59c49c120b5117cd8/pkg/jobs/jobspb/schedule.proto#L53)):*
> ```protobuf
> message ExecutionArguments {
>   google.protobuf.Any args = 1;
>   string create_statement = 2;
> ```

nit: create_schedule_statement

And a comment: // String representation of the query used to create the schedule.
___
*[pkg/sql/parser/sql.y, line 5388 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mcy8wc5Clpv0_zbEDia:-Mcy8wc5Clpv0_zbEDib:b-irkyz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6dfc3bd22b66710435adab59c49c120b5117cd8/pkg/sql/parser/sql.y#L5388)):*
> ```Yacc
> | SHOW CREATE error // SHOW HELP: SHOW CREATE
> 
> // %Help: SHOW CREATE SCHEDULE - display the CREATE statement for a scheduled BACKUP
> ```

nit: scheduled job
___
*[pkg/sql/sem/tree/show.go, line 533 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mcy90ZdA7sjH7afKcd1:-Mcy90ZdA7sjH7afKcd2:brllx9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6dfc3bd22b66710435adab59c49c120b5117cd8/pkg/sql/sem/tree/show.go#L533)):*
> ```Go
> 
> // ShowCreateSchedules represents a SHOW CREATE SCHEDULE statement.
> type ShowCreateSchedules struct {
> ```

nit: rename to ShowCreateSchedule
___
*[pkg/sql/sqltelemetry/show.go, line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-McyA1EvCBtoPq4nvVwt:-McyA1EvCBtoPq4nvVwu:b20kpl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6dfc3bd22b66710435adab59c49c120b5117cd8/pkg/sql/sqltelemetry/show.go#L68)):*
> ```Go
> 	Locality:                "locality",
> 	Create:                  "create",
> 	CreateSchedule:          "create schedule",
> ```

might need an _ here

**Comment by adityamaru** [comment]
Can we add a few simple sub-tests for this? Maybe a `create schedule` with different kinds of options, and then compare the string you passed into the returned query? I'm interested in seeing RECURRING set to a cron format.
Also, I believe this should be fine out of the box, but can you ensure that explicit AUTH access keys are sanitized when we display this statement?

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @annezhu98, and @miretskiy)

___
*[pkg/ccl/backupccl/backup.proto, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mcu4t5i2_qX8OtSR0Q8:-McyFXwG0mFjnEY1Wfg_:b-h2kz4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9161773c608b8639880a8f8eeba7f3f84f48421d/pkg/ccl/backupccl/backup.proto#L160)):*
<details><summary><i>Previously, annezhu98 (Anne Zhu) wrote…</i></summary><blockquote>

I believe this would be specific to scheduled BACKUPs, similar to `SHOW SCHEDULE`
</blockquote></details>

Sure, but the command you're implementing is not `SHOW BACKUP SCHEDULE`.  You're implementing `SHOW CREATE SCHEDULE` --the command used to create the schedule. Schedules are not specific to backups.
 As such, I think either the option has to be schedule specific, or backup specific -- but in this case the command must reflect that.

Another question is: Is the purpose of this string representation to be able to recreate command as used when creating the schedule *exactly*? Or is it okay to re-synthesize the create command?

I'm actually not sure at all that the exact command is required here.  One possibility is that instead of this command using delegator pattern, we implement `planNode` interface -- see `control_schedules.go`.
Basically, for each of the show schedules, you would load that schedule, you would get an executor for that schedule; and from there you should be able to synthesize correct `tree.ScheduledBackup` node -- and then just call tree.AsString on that.

I honestly don't think you need to change any protos; and you can do this in a way that's generic to all schedules, not just backup schedules.

**Review by miretskiy — CHANGES_REQUESTED** [review]
Very good progress. I still have some comments and questions, but I think this PR is getting close.

> Reviewed 1 of 1 files at r2, 13 of 15 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-:-MdUeCgs4Zwph9loYiON:b7sj1ks)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annezhu98)

___
*[pkg/ccl/backupccl/backup.proto, line 160 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU6fhQ89PAdjWWuX5e:-MdU6fhQ89PAdjWWuX5f:bbmqf0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/ccl/backupccl/backup.proto#L160)):*
> ```protobuf
>   int64 unpause_on_success = 3;
>   bool updates_last_backup_metric = 4;
>   string dependent_schedule_recurrence = 5;
> ```

I know that this is the only piece of information you need right now to reconstruct the scheduled, but I'd much rather store dependent schedule ID.  Perhaps you'd want to load more information eventually.  And it's just strange to see some random string -- vs ID.

Set this field to this schedule ID if there is no dependent schedule.  
___
*[pkg/ccl/backupccl/schedule_exec.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU7VuaDtIFvtzQt4FN:-MdU7VuaDtIFvtzQt4FO:b-p34qnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/ccl/backupccl/schedule_exec.go#L174)):*
> ```Go
> }
> 
> func (e *scheduledBackupExecutor) CreateString(sj *jobs.ScheduledJob) (string, error) {
> ```

This is exactly how I envisioned this would work.  Very nice -- good job.
___
*[pkg/jobs/executor_impl.go, line 96 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU7kacAydN__qkinzK:-MdU7kacAydN__qkinzL:b8yw19b) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/jobs/executor_impl.go#L96)):*
> ```Go
> 
> func (e *inlineScheduledJobExecutor) CreateString(_ *ScheduledJob) (string, error) {
> 	return "", nil
> ```

just return
`"", errors.AssertionFailedf("unimplemented")`
___
*[pkg/jobs/scheduled_job_executor.go, line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU5SAg3rydPCo-LS2o:-MdU5SAhBm5m29nl7_tj:b-2dc2dz) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/jobs/scheduled_job_executor.go#L54)):*
> ```Go
> 	Metrics() metric.Struct
> 
> 	// CreateString returns the create statement used to execute the scheduled job.
> ```

Perhaps rename to CreateScheduleString? Or maybe even `CreateScheduleStatementString`?
Also, comment should be updated.  Something like
```
// CreateScheduleString returns the statement used to create schedule for the scheduled job executed
// by this executor?
```
___
*[pkg/jobs/jobspb/schedule.proto, line 48 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU82YC30e5dtiKJHKF:-MdU82YC30e5dtiKJHKG:b9ck0e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/jobs/jobspb/schedule.proto#L48)):*
> ```protobuf
>   // How to handle failed jobs.
>   ErrorHandlingBehavior on_error = 2;
>   bool ignore_existing_backups = 3;
> ```

I don't think we need to add this field at all.  This option is used only when creating the schedule and is checked only once.  

Presumably, the users want to see the schedule creating statement for either debugging purpose, or perhaps they want to "copy & paste" to create new schedule.  In that case, if they make a mistake and not update the schedule appropriately (e.g. by keeping the same path), I'd rather return an error and require this option to be specified explicitly.
___
*[pkg/sql/control_schedules.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU5orW1RqzfEGHNeYr:-MdU5orW1RqzfEGHNeYs:big6rdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/sql/control_schedules.go#L72)):*
> ```Go
> 		fmt.Sprintf(
> 			"SELECT schedule_id, schedule_name, schedule_expr, schedule_details, " +
> 				"executor_type, execution_args FROM %s WHERE schedule_id = $1",
> ```

probably could be changed to `SELECT *` at this point.
___
*[pkg/sql/show_create_schedule.go, line 48 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU9xTc3Ptpydd48deK:-MdU9xTc3Ptpydd48deL:bhyakb) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/sql/show_create_schedule.go#L48)):*
> ```Go
> 			sjIdDatum := tree.NewDInt(tree.DInt(sjID))
> 			sj, err := loadSchedule(runParams{ctx: ctx, p: p,
> 				extendedEvalCtx: &p.extendedEvalCtx}, sjIdDatum)
> ```

On the one hand, it's nice that this statement is in its own file.  On the other hand, I really
don't like the fact that we calling loadSchedule defined in the control_schedules.go

I understand why you did this though -- control_schedules returns fixed set of columns.
But I wonder if we can perhaps change control schedules to use delayedNode somehow...

Anyways, I also noticed that because you're not using control schedules, you also seem to be missing
admin check for this statement.
___
*[pkg/sql/parser/sql.y, line 5399 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU93HPEtHzSuUV2l8l:-MdU93HPEtHzSuUV2l8m:b-f4wqgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/sql/parser/sql.y#L5399)):*
> ```Yacc
>     $$.val = &tree.ShowCreateSchedules{ScheduleID: $4.expr()}
>   }
> | SHOW CREATE SCHEDULE error // SHOW HELP: SHOW CREATE SCHEDULE
> ```

All of the other job control statements support plurals (`PAUSE JOBS`, `RESUME SCHEDULES`, etc).  We should probably do the same.

**Review by miretskiy — COMMENTED** [review]
> Reviewed 1 of 15 files at r5, 6 of 14 files at r6, 4 of 12 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @annezhu98, and @miretskiy)

___
*[pkg/sql/show_create_schedule.go, line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Me7SxoGCoUogtOd4-en:-Me7SxoGCoUogtOd4-eo:b-rwnea7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6f68cc6f0fe36701c57c95df90375453d88c300/pkg/sql/show_create_schedule.go#L67)):*
> ```Go
> 			"load-schedules",
> 			params.EvalContext().Txn, sessiondata.InternalExecutorOverride{User: security.RootUserName()},
> 			fmt.Sprintf("SELECT * FROM %s", env.ScheduledJobsTableName()))
> ```

It seems to me that you can save quite a bit of typing if you build the query you want to execute as you go along... Something like this:

```
query := fmt.Sprintf("SELECT * FROM %s", env.ScheduledJobsTableName())
whereClause := ""
var qArgs  []interface{}

if n.ScheduleID != nil {
   ...
   qArgs = append(qArgs, tree.NewDInt(tree.DInt(sjID))
   whereClause = "WHERE schedule_id = $1"
}
// Execute query
datums, columns, err := .... Query(...., fmt.Sprintf("%s %s", query, whereClause),  qArgs...)
```

**Review by miretskiy — APPROVED** [review]
> Reviewed 1 of 12 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @annezhu98, and @miretskiy)

**Review by annezhu98 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @miretskiy)

___
*[pkg/ccl/backupccl/backup.proto, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mcu4t5i2_qX8OtSR0Q8:-Mdx4zi7D5xQFRieGUcE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9161773c608b8639880a8f8eeba7f3f84f48421d/pkg/ccl/backupccl/backup.proto#L160)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Sure, but the command you're implementing is not `SHOW BACKUP SCHEDULE`.  You're implementing `SHOW CREATE SCHEDULE` --the command used to create the schedule. Schedules are not specific to backups.
 As such, I think either the option has to be schedule specific, or backup specific -- but in this case the command must reflect that.

Another question is: Is the purpose of this string representation to be able to recreate command as used when creating the schedule *exactly*? Or is it okay to re-synthesize the create command?

I'm actually not sure at all that the exact command is required here.  One possibility is that instead of this command using delegator pattern, we implement `planNode` interface -- see `control_schedules.go`.
Basically, for each of the show schedules, you would load that schedule, you would get an executor for that schedule; and from there you should be able to synthesize correct `tree.ScheduledBackup` node -- and then just call tree.AsString on that.

I honestly don't think you need to change any protos; and you can do this in a way that's generic to all schedules, not just backup schedules.
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup.proto, line 160 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU6fhQ89PAdjWWuX5e:-MeC8xV915RWpSYZib9A:b-qtn34i) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/ccl/backupccl/backup.proto#L160)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I know that this is the only piece of information you need right now to reconstruct the scheduled, but I'd much rather store dependent schedule ID.  Perhaps you'd want to load more information eventually.  And it's just strange to see some random string -- vs ID.

Set this field to this schedule ID if there is no dependent schedule.  
</blockquote></details>

As discussed offline, storing the schedule crontab seems to be a better option since we would avoid doing an extra query to fetch the crontab of the dependent full backup.
___
*[pkg/ccl/backupccl/schedule_exec.go, line 174 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU7VuaDtIFvtzQt4FN:-Mdx5Ws3-Ho9FsWfyRnw:b75099u) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/ccl/backupccl/schedule_exec.go#L174)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This is exactly how I envisioned this would work.  Very nice -- good job.
</blockquote></details>

Thank you! :)
___
*[pkg/jobs/executor_impl.go, line 96 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU7kacAydN__qkinzK:-MdZtU_O4mNldCTV0cRk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/jobs/executor_impl.go#L96)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

just return
`"", errors.AssertionFailedf("unimplemented")`

</blockquote></details>

Done.
___
*[pkg/jobs/scheduled_job_executor.go, line 54 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU5SAg3rydPCo-LS2o:-MdZtTg0CjMfsrIvrFun:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/jobs/scheduled_job_executor.go#L54)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Perhaps rename to CreateScheduleString? Or maybe even `CreateScheduleStatementString`?
Also, comment should be updated.  Something like
```
// CreateScheduleString returns the statement used to create schedule for the scheduled job executed
// by this executor?
```
</blockquote></details>

Done.
___
*[pkg/sql/show_create_schedule.go, line 48 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU9xTc3Ptpydd48deK:-Mdx4hdx-Z1J4w6D5R83:b-ukcuqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/sql/show_create_schedule.go#L48)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

On the one hand, it's nice that this statement is in its own file.  On the other hand, I really
don't like the fact that we calling loadSchedule defined in the control_schedules.go

I understand why you did this though -- control_schedules returns fixed set of columns.
But I wonder if we can perhaps change control schedules to use delayedNode somehow...

Anyways, I also noticed that because you're not using control schedules, you also seem to be missing
admin check for this statement.
</blockquote></details>

Done. As for `control_schedules`, I think implementing `planNodeFastPath` makes sense since all we need to return is the number of rows affected. 
___
*[pkg/sql/show_create_schedule.go, line 67 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Me7SxoGCoUogtOd4-en:-MeC81qYEVllbwUbIrgK:b-2we8ib) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6f68cc6f0fe36701c57c95df90375453d88c300/pkg/sql/show_create_schedule.go#L67)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

It seems to me that you can save quite a bit of typing if you build the query you want to execute as you go along... Something like this:

```
query := fmt.Sprintf("SELECT * FROM %s", env.ScheduledJobsTableName())
whereClause := ""
var qArgs  []interface{}

if n.ScheduleID != nil {
   ...
   qArgs = append(qArgs, tree.NewDInt(tree.DInt(sjID))
   whereClause = "WHERE schedule_id = $1"
}
// Execute query
datums, columns, err := .... Query(...., fmt.Sprintf("%s %s", query, whereClause),  qArgs...)
```
</blockquote></details>

Kept this as is since the executors used for the two queries are different, so we will end up doing pretty much the same thing.
___
*[pkg/sql/parser/sql.y, line 5399 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU93HPEtHzSuUV2l8l:-MeC8aPbClDmPmE1wznN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/sql/parser/sql.y#L5399)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

All of the other job control statements support plurals (`PAUSE JOBS`, `RESUME SCHEDULES`, etc).  We should probably do the same.
</blockquote></details>

Done.
___
*[pkg/sql/sqltelemetry/show.go, line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-McyA1EvCBtoPq4nvVwt:-MdG-E2L0fyCVmeZkixU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6dfc3bd22b66710435adab59c49c120b5117cd8/pkg/sql/sqltelemetry/show.go#L68)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

might need an _ here
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/schedule.proto, line 48 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU82YC30e5dtiKJHKF:-MdZuQ3r3NZ8lfeP84Fe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/jobs/jobspb/schedule.proto#L48)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't think we need to add this field at all.  This option is used only when creating the schedule and is checked only once.  

Presumably, the users want to see the schedule creating statement for either debugging purpose, or perhaps they want to "copy & paste" to create new schedule.  In that case, if they make a mistake and not update the schedule appropriately (e.g. by keeping the same path), I'd rather return an error and require this option to be specified explicitly.
</blockquote></details>

Done.
___
*[pkg/sql/control_schedules.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MdU5orW1RqzfEGHNeYr:-Mdx5_ZEAQrtmucp3oPo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/470da3b502c6f6b5f01f23d02f7b4c0663491758/pkg/sql/control_schedules.go#L72)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

probably could be changed to `SELECT *` at this point.
</blockquote></details>

Done.

**Review by adityamaru — CHANGES_REQUESTED** [review]
> Reviewed 3 of 14 files at r6, 4 of 12 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annezhu98 and @miretskiy)

___
*[pkg/ccl/backupccl/backup.proto, line 160 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Meze0me1GD8yroY3OMt:-Meze0me1GD8yroY3OMu:b1bzov8) ([raw file](https://github.com/cockroachdb/cockroach/blob/12e703cdc6b0dc3e08015d809e1eff803268c76e/pkg/ccl/backupccl/backup.proto#L160)):*
> ```protobuf
>   int64 unpause_on_success = 3;
>   bool updates_last_backup_metric = 4;
>   string dependent_schedule_crontab = 5;
> ```

can we add a comment here describing what this field represents?
___
*[pkg/jobs/scheduled_job_executor_test.go, line 112 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mf-uxB31AQdmSFDakH5:-Mf-uxB31AQdmSFDakH6:b-t04a9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/129765d45dd6a2c8c2d8273956d0b56292e11ca1/pkg/jobs/scheduled_job_executor_test.go#L112)):*
> ```Go
> }
> 
> func TestCreateScheduledJobStatement(t *testing.T) {
> ```

hmmm, im not sure this test is very useful. It just seems to be checking that the test executor does not have the method implemented?
___
*[pkg/sql/parser/sql.y, line 4650 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MezU7pw4De7iY6rD-wD:-MezU7pxArsMh1ItVnhH:b-xwrtgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/12e703cdc6b0dc3e08015d809e1eff803268c76e/pkg/sql/parser/sql.y#L4650)):*
> ```Yacc
> | show_constraints_stmt     // EXTEND WITH HELP: SHOW CONSTRAINTS
> | show_create_stmt          // EXTEND WITH HELP: SHOW CREATE
> | show_create_schedules_stmt // EXTEND WITH HELP: SHOW CREATE SCHEDULES
> ```

nit: fix comment spacing
___
*[pkg/sql/parser/sql.y, line 5395 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MezYvMiD1D6wSuEI91_:-MezYvMiD1D6wSuEI91a:boldv0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/12e703cdc6b0dc3e08015d809e1eff803268c76e/pkg/sql/parser/sql.y#L5395)):*
> ```
> SHOW
> ```
nit: fix spacing
___
*[pkg/sql/parser/sql.y, line 5403 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-MezU5DqB66iAqLCFoyA:-MezU5DqB66iAqLCFoyB:bc8r2la) ([raw file](https://github.com/cockroachdb/cockroach/blob/12e703cdc6b0dc3e08015d809e1eff803268c76e/pkg/sql/parser/sql.y#L5403)):*
> ```Yacc
> | SHOW CREATE SCHEDULE a_expr
>   {
>   	/* SKIP DOC */
> ```

nit: fix spacing

**Comment by adityamaru** [comment]
Thanks for working through this @annezhu98. We're almost there! Can you update the commit message to reflect what we had discussed offline about `SHOW CREATE SCHEDULE` when run against the full/incremental schedules implicitly created for a single `CREATE SCHEDULE`, return the original `CREATE SCHEDULE` statement.

Is it possible to add a test for `SHOW CREATE SCHEDULES`?

**Review by annezhu98 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66782)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @miretskiy)

___
*[pkg/ccl/backupccl/backup.proto, line 160 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Meze0me1GD8yroY3OMt:-Mf3baMh1u25JBQUdYkS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12e703cdc6b0dc3e08015d809e1eff803268c76e/pkg/ccl/backupccl/backup.proto#L160)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

can we add a comment here describing what this field represents?
</blockquote></details>

Done.
___
*[pkg/jobs/scheduled_job_executor_test.go, line 112 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66782#-Mf-uxB31AQdmSFDakH5:-Mf4IdFE3lDzTtYa53Kp:b-p4j7g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/129765d45dd6a2c8c2d8273956d0b56292e11ca1/pkg/jobs/scheduled_job_executor_test.go#L112)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

hmmm, im not sure this test is very useful. It just seems to be checking that the test executor does not have the method implemented?
</blockquote></details>

Removed the test, it makes more sense for each executor to test its own implementation of the method separately.

**Comment by adityamaru** [comment]
LGTM!! Once CI passes feel free to bors it 🙂
